cloudinit/config/cc_ansible.py

```diff
@@ -0,0 +1,116 @@
+from cloudinit.subp import ProcessExecutionError, subp, which
+
+meta: MetaSchema = {
+    "id": "cc_ansible",
+    "name": "Ansible",
+    "title": "Configure ansible for instance",
+    "description": dedent(
+        """\
+        This module provides ansible integration.``
+
+        Ansible is often used agentless and in parallel
+        across multiple hosts simultaneously. This
+        doesn't fit the model of cloud-init: a single
+        host configuring itself during boot. Instead,
+        this module installs ansible during boot and
+        then uses ``ansible-pull`` to run the playbook
+        repository at the remote URL.
+        """
+    ),
+    "distros": [ALL_DISTROS],
+    "examples": [
+        dedent(
+            """\
+            #cloud-config
+            ansible:
+              install: true
+              pull:
+                url: "https://github.com/holmanb/vmboot.git"
+                playbook-name: ubuntu.yml
+            """
+        ),
+    ],
+    "frequency": PER_INSTANCE,
+}
+
+__doc__ = get_meta_doc(meta)
+
+
+def handle(name: str, cfg: dict, cloud: Cloud, log: Logger, _):
+
+    # TODO: Remove this before PR
+    if not all([name, cfg, cloud, log]):
+        raise ValueError(
+            f"Configuration not supported: {name} {cfg} {cloud} {log}"
+        )
+
+    install, ansible_config = get_and_validate_config(cfg)
+    install_ansible(cloud.distro, install)
+    run_ansible_pull(ansible_config, log)
+
+
+def get_and_validate_config(cfg: dict) -> Tuple[bool, dict]:
+    ansible_cfg: dict = cfg.get("ansible", {})
+    pull: dict = ansible_cfg.get("pull", {})
+    install: bool = ansible_cfg.get("install", False)
+    if pull:
+        return (
+            install,
+            pull,
+        )
+    raise ValueError("ansible module key missing (requires 'pull')")
+
+
+def install_ansible(distro: Distro, install: bool):
+    """Give users flexibility in whether to use the package install module or
+    this module.
+    """
+    if install:
+        distro.install_packages("ansible")
+
+
+def check_deps(dep: str):
+    if not which(dep):
+        raise ValueError(
+            f"command: {dep} is not available, please set"
+            "ansible.install: True in your config or ensure that"
+            "it is installed (either in your base image or in a package"
+            "install module)"
+        )
+
+
+def run_ansible_pull(cfg: dict, log: Logger):
+    cmd = "ansible-pull"
+    check_deps(cmd)
+    try:
+        playbook_name = cfg.get("playbook-name")
+        if not playbook_name:
+            raise ValueError("Missing required key: 'playbook-name'")
+
+        stdout, stderr = subp(
+            [
+                cmd,
+                *[
+                    f"--{key}={value}" if value else f"--{key}"
+                    for key, value in cfg.items()
+                    if key != "playbook-name"
+                ],
+                playbook_name,
+            ]
+        )
+        if stderr:
```

Comment:
These logs would be really helpful for debugging, but maybe `/var/log/cloud-init-output.log` would be a better place for them?